linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IntegerDivisions.scala

```diff
@@ -0,0 +1,693 @@
+        quotient
+      } else {
+        // n - divisor * quotient
+        BinaryOp(int.Op_-, argRef, BinaryOp(int.Op_*, int.literal(divisor), quotient))
+      }
+    }
+  }
+
+  private def argRef[T](implicit int: UnsignedIntegral[T], pos: Position): VarRef =
+    VarRef(NumeratorArgName)(int.tpe)
+
+  private def genZero()(implicit pos: Position): Tree = {
+    UnaryOp(
+      UnaryOp.Throw,
+      New(
+        ArithmeticExceptionClass,
+        MethodIdent(MethodName.constructor(List(ClassRef(BoxedStringClass)))),
+        List(StringLiteral("/ by zero"))
+      )
+    )
+  }
+
+  private def genOne[T](
+      isQuotient: Boolean, negativeDivisor: Boolean)(
+      implicit int: UnsignedIntegral[T], pos: Position): Tree = {
+
+    // Straightforward
+
+    if (isQuotient) {
+      if (negativeDivisor)
+        BinaryOp(int.Op_-, int.literalZero, argRef)
+      else
+        argRef
+    } else {
+      int.literalZero
+    }
+  }
+
+  private def genPowerOfTwoSigned[T](
+      isQuotient: Boolean, negativeDivisor: Boolean, absDivisor: T)(
+      implicit int: UnsignedIntegral[T], pos: Position): Tree = {
+
+    // Sections 10-1 (quotient) and 10-2 (remainder)
+
+    val k = int.log2Exact(absDivisor) // i.e., absDivisor == 2^k
+
+    var t = BinaryOp(int.Op_>>, argRef, IntLiteral(k - 1))
+    t = BinaryOp(int.Op_>>>, t, IntLiteral(int.bitSize - k))
+
+    if (isQuotient) {
+      t = BinaryOp(int.Op_+, argRef, t)
+      val q = BinaryOp(int.Op_>>, t, IntLiteral(k))
+
+      if (negativeDivisor)
+        BinaryOp(int.Op_-, int.literalZero, q)
+      else
+        q
+    } else {
+      val temp1Def = tempVarDef(tName, t)
+      Block(
+        temp1Def,
+        // ((arg + temp1) & (2^k - 1)) - temp1
+        BinaryOp(
+          int.Op_-,
+          BinaryOp(
+            int.Op_&,
+            BinaryOp(int.Op_+, argRef, temp1Def.ref),
+            int.literal(int.minus(absDivisor, int.one))
+          ),
+          temp1Def.ref
+        )
+      )
+    }
+  }
+
+  private def genPowerOfTwoUnsigned[T](
+      isQuotient: Boolean, absDivisor: T)(
+      implicit int: UnsignedIntegral[T], pos: Position): Tree = {
+
+    // Straightforward
+
+    val k = int.log2Exact(absDivisor) // i.e., absDivisor == 2^k
+
+    if (isQuotient)
+      BinaryOp(int.Op_>>>, argRef, IntLiteral(k))
+    else
+      BinaryOp(int.Op_&, argRef, int.literal(int.minus(absDivisor, int.one)))
+  }
+
+  private def genNonPowerOfTwoSignedQuotient[T](
+      negativeDivisor: Boolean, absDivisor: T)(
+      implicit int: UnsignedIntegral[T], pos: Position): Tree = {
+
+    // Hacker's Delight, Sections 10-3 to 10-6
+
+    val data = computeSignedMagic(absDivisor, negativeDivisor)
+
+    var q = int.genMulSignedHi(data.M, argRef, useRuntimeLong)
+    if (data.add > 0)
+      q = BinaryOp(int.Op_+, q, argRef)
+    else if (data.add < 0)
+      q = BinaryOp(int.Op_-, q, argRef) // Section 10-5 for negative divisors
+    q = BinaryOp(int.Op_>>, q, IntLiteral(data.shift))
```

Comment:
Note to self (check later): if shift is 0, does this disappear / can shift be 0?